### ISSUE TITLE
There seems to be a mistake in `FindPath`

### ISSUE THREAD

**Issue by dylanwilson80** [adammyhre/Unity-GOAP#3]
First of all let me just say I really appreciate the video. It was a great starting point for my GOAP implementation and helped me wrap my head around the algorithm. I really like your simple take on it. 

That said, I was trying your GOAP implementation in my project and I started noticing an issue. Sometimes the plan would spit out an action that didn't make sense. At first I thought it might've been just because my use case was a bit different. I had already refactored the code quite a bit to suit my game so I wasn't sure.

I ignored the issue at first because I had other things I needed to work on. Then eventually I got around to writing some unit tests around the planner. This is the test I came up with...

```
        var gameObject = new GameObject();
        var npc = new TestNpc();
        var target = gameObject.AddComponent<GoapTarget>();
        
        var playerDead = GoapBelief.Create("Player Dead", () => false);
        var hasNoWeapon = GoapBelief.Create("Has No Weapon", () => true);
        var hasWeapon = GoapBelief.Create("Has Weapon", () => false);
        
        var goals = new HashSet<GoapGoal> { new("Kill Player") { DesiredEffects = { playerDead } } };
        var findWeapon = new GoapAction("Find Weapon", npc, target) { Preconditions = { hasNoWeapon }, Effects = { hasWeapon } };
        var attackPlayer = new GoapAction("Attack Player", npc, target) { Preconditions = { hasWeapon }, Effects = { playerDead } };
        var actions = new HashSet<GoapAction> { findWeapon, attackPlayer };
        yield return null;
        
        var planner = new GoapPlanner();
        var plan = planner.CreatePlan(goals, actions);
        
        Assert.IsNotNull(plan);
        Assert.AreEqual(findWeapon, plan.Actions.First());
```

My code as deviated a bit from your original design but I think you get the idea. Basically in this very simple scenario which I would describe as "Enemy needs weapon to attack" it wouldn't ever take the "Find weapon" action first.

The bug, it appears, is found in the `FindPath` method. Right after creating the `newAvailableActions`. In the original implementation when creating a new node in the tree. 

```
                var newAvailableActions = new HashSet<AgentAction>(actions);
                newAvailableActions.Remove(action);
                
                var newNode = new Node(parent, action, newRequiredEffects, parent.Cost + action.Cost);
                
                // Explore the new node recursively
                if (FindPath(newNode, newAvailableActions)) {
                    parent.Leaves.Add(newNode);
                    newRequiredEffects.ExceptWith(newNode.Action.Preconditions);
                }
```

This code looks correct at first glance because you're passing the `newAvailableActions` into the recursive call to `FindPath` but this is misleading because as far as I can tell `newNode` is supposed to represent the precondition to the current action.

In other words, in this scenario the action we're trying to take is "Attack Player" which has a precondition of "Find Weapon". At this point in the code the `action` variable is "Attack Player" and the `newAvailableActions` is the list of preconditions `[ "Find Weapon" ]`. So it seems like the correct thing to do (unless I've misunderstood something) is to recursively explore the "Find Weapon" node to see if that is the next action we can take. 

Here's the implementation I came up with that passes my test.

```
                var newAvailableActions = new HashSet<GoapAction>(availableActions);
                newAvailableActions.Remove(action);

                foreach (var subAction in newAvailableActions)
                {
                    var newGoal = new GoapNode(goalNode, subAction, newDesiredEffects, goalNode.Cost + action.Cost);
                    
                    if (FindPath(newGoal, newAvailableActions))
                        goalNode.Leaves.Add(new GoapNode(newGoal, subAction, subAction.Preconditions, goalNode.Cost));
                }
```

Now to be clear, I haven't tested many other scenarios yet. This implementation is currently working in my game but there might be cases I've overlooked at this point. And to be honest, now that I'm getting my head around how the algorithm works and getting more tests in place I'm probably going to rewrite the whole thing from scratch.

**Comment by goncastrum** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by adammyhre** [comment]
Thank you for taking the time to provide such a detailed explanation and test case! After review, I can see that the FindPath method can possibly be improved to handle action dependencies better, especially in cases where multiple prerequisites need to be fulfilled in a specific order.  I'll probably be making a small adjustment to the current implementation after more testing to better handle these dependencies as you've suggested. A change should help improve the planner's behavior in scenarios like the one you described.

That said, as noted in the code comments and the video, a more powerful search algorithm, such as A* or D*, would be a more robust solution for planning scenarios. These algorithms would provide more efficient pathfinding and better handling of complex goal dependencies than a basic DFS.

**Comment by dylanwilson80** [comment]
To be clear, the only reason I commented on YouTube and raised the issue here on GitHub was to help other people understand the limitations of this implementation before using it. 

If you're just going to pre-emptively close the issue and not do anything about it that somewhat defeats the purpose and makes me feel like I wasted several hours finding, understanding and documenting the bug for no reason.

More importantly, closing the issue makes it appear like there are zero problems in this repository and will likely waste someone else's time in the future when they run into the same bug in their own games.
